src/exec/use_pty/pipe.rs

```diff
@@ -103,23 +103,37 @@ impl<L: Read + Write + AsRawFd, R: Read + Write + AsRawFd> Pipe<L, R> {
     }
 }
 
-/// The size of the internal buffer of the pipe.
-const BUFSIZE: usize = 6 * 1024;
-
-/// A buffer that stores the bytes read from `R` before they are written to `W`.
+/// A circular buffer that stores the bytes read from `R` before they are written to `W`.
+///
 struct Buffer<R, W> {
-    buffer: [u8; BUFSIZE],
-    /// The start of the busy section of the buffer.
-    start: usize,
-    /// The end of the busy section of the buffer.
-    end: usize,
+    /// The internal buffer.
+    /// ```text
+    ///
+    /// ┌───┬───┬───┬───┬───┬───┬───┬───┐
+    /// │   │ M │ D │ D │ D │   │   │   │
+    /// └───┴───┴───┴───┴───┴───┴───┴───┘
+    ///           ▲           ▲
+    ///           │           │
+    ///         head         tail
+    /// ```
+    /// The extra byte (`M` in the diagram) is used as a marker to differentiate an empty buffer
+    /// from a full one and it is always located one position before head.
+    buffer: [u8; MAX_POS + 1],
+    /// The first location of the buffer that has data.
+    head: usize,
+    /// The next location at which new data will be inserted.
+    tail: usize,
     /// The handle for the event of the reader.
     read_handle: EventHandle,
     /// The handle for the event of the writer.
     write_handle: EventHandle,
+    /// A type marker so we don't read or write from the wrong end.
     marker: PhantomData<(R, W)>,
 }
 
+/// The maximum value that head and tail can take.
+const MAX_POS: usize = 6 * 1024 - 1;
```

Comment:
nit: this isn't an embedded device so neither the performance or binary size of this abstraction matter in the great scheme of things but `% power_of_two` is faster than `% non_power_of_two` because the former is optimized to a bitmask (`& (1 << exponent) - 1`)

Reply:
oh, but we never do `x % MAX_POS`, just `x % (MAX_POS + 1)` which is a power of two as you mentioned

Reply:
right, but **6** times 1024 is not a power of 2. `8 * 1024` is; could that be a typo?